seisbench/generate/augmentation.py

```diff
@@ -670,3 +672,210 @@ def __call__(self, state_dict):
+                                adding noise samples, if e.g. the signal-to-noise ratio is below a certain threshold:
+                                ``metadata_thresholds={"trace_Z_snr_db": 10}``
+                                In the example above, noise is only added when SNR of signal of the Z-component is equal
+                                or greater than 10.
+                                Default is None
+    """
+
+    def __init__(
+        self,
+        noise_dataset: seisbench.data.base.WaveformDataset,
+        scale: tuple = (0, 1),
+        scaling_type: str = "peak",
+        metadata_thresholds: (dict, None) = None,
+        key: str = "X",
+    ):
+        if isinstance(key, str):
+            self.key = (key, key)
+        else:
+            self.key = key
+
+        if scaling_type.lower() not in ["std", "peak"]:
+            msg = "Argument scaling_type must be either 'std' or 'peak'."
+            raise ValueError(msg)
+
+        self.noise_dataset = noise_dataset
+        self.scale = scale
+        self.scaling_type = scaling_type.lower()
+        self.noise_generator = seisbench.generate.GenericGenerator(noise_dataset)
+        self.metadata_thresholds = metadata_thresholds
+        self.noise_samples = len(noise_dataset)
+
+    def __call__(self, state_dict):
+        x, metadata = state_dict[self.key[0]]
+
+        if self.key[0] != self.key[1]:
+            # Ensure data and metadata is not modified inplace unless input and output key are anyhow identical
+            metadata = copy.deepcopy(metadata)
+            x = x.copy()
+
+        # Check all requirements from metadata thresholds
+        # If value from metadata is below given threshold value, then x is returned without added noise
+        if self.metadata_thresholds:
+            for threshold_key, item in self.metadata_thresholds.items():
+                if metadata[threshold_key] < item:
+                    state_dict[self.key[1]] = (x, metadata)
+                    return
+
+        # Defining scale for noise amplitude
+        scale = np.random.uniform(*self.scale)
+
+        # Removing mean from data x
+        x -= np.mean(x, axis=-1, keepdims=True)
+
+        if self.scaling_type == "peak":
+            scale = scale * np.max(np.abs(x))
+        elif self.scaling_type == "std":
+            scale = scale * np.std(x)
+
+        # Draw random noise sample from the dataset and cut to same length as x
+        n = self.noise_generator[np.random.randint(low=0, high=self.noise_samples)]["X"]
+
+        # Removing mean from noise samples
+        n -= np.mean(n, axis=-1, keepdims=True)
+
+        # Normalize noise samples
+        if self.scaling_type == "peak":
+            n = n / np.max(np.abs(n))
+        elif self.scaling_type == "std":
+            n = n / np.std(n)
+        n = n * scale
+
+        # Cutting noise to same length as x
+        if n.shape[1] - x.shape[1] - 1 < 0:
```

Comment:
I'm not sure I understand why there is a `-1` here. From how I understand this it should not fail for `n.shape[1] == x.shape[1]`, but it currently does.